File: go/day5_part1_2017.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	// Step 1: Read Input
	data, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Println("File reading error", err)
		return
	}

	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	var offsets []int
	for _, line := range lines {
		offset, _ := strconv.Atoi(line)
		offsets = append(offsets, offset)
	}

	// Step 2: Initialize Variables
	index := 0
	steps := 0

	// Step 3: Navigate Maze
	for index >= 0 && index < len(offsets) {
		// Fetch the jump offset at the current index
		jump := offsets[index]

		// Step 4: Update Offset
		offsets[index]++

		// Move to the new index
		index += jump

		// Increment steps counter
		steps++
	}

	// Step 6: Output
	fmt.Println(steps)
}

File: nim/day5_part1_2017.nim

import os
import strutils

var offsets: seq[int]

let data = readFile("input.txt")
for line in data.splitLines():
  offsets.add(parseInt(line))

var index = 0
var steps = 0

while index >= 0 and index < offsets.len:
  let jump = offsets[index]
  offsets[index] += 1
  index += jump
  inc(steps)

echo steps
